### PR TITLE
Thumbnails

### LINKED ISSUE #223: [Feature]: add precise seeking with thumbnails
### What feature do you want to add?

precisee seeking with thumbnails

### Why do you want to have this feature?

its a indispensable feature

### Anything other details to share?

_No response_

### PR DESCRIPTION
**Pull request by frost768**
resolves #223 
Preview:

https://github.com/movie-web/movie-web/assets/19949367/9e607bf7-970f-4962-aae8-46f2e1032818

### CODE REVIEW

File: src/utils/thumbnailCreator.ts [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+): AsyncGenerator<Thumbnail, Thumbnail> {
+  const canvas = document.createElement("canvas");
+  const ctx = canvas.getContext("2d");
+  if (!ctx) return { from: -1, to: -1, imgUrl: "" };
+  const video = document.createElement("video");
+  video.src = videoUrl;
+  video.crossOrigin = "anonymous";
+
+  // Wait for the video metadata to load
+  const metadata = await new Promise((resolve, reject) => {
```
Comment (thehairy) [inline src/utils/thumbnailCreator.ts:19]:
```suggestion
  await new Promise((resolve, reject) => {
```

metadata is not used, linter is complaining.

**Review by thehairy — CHANGES_REQUESTED** [review]

**Review by thehairy — APPROVED** [review]

File: src/utils/thumbnailCreator.ts [hunk 2]
```diff
@@ -0,0 +1,51 @@
+export interface Thumbnail {
+  from: number;
+  to: number;
+  imgUrl: string;
+}
+export const SCALE_FACTOR = 1;
+export default async function* extractThumbnails(
```
Comment (mrjvs) [inline src/utils/thumbnailCreator.ts:7]: first time I've ever seen a generator function used in the wild, exciting

Comment (frost768) [reply src/utils/thumbnailCreator.ts:7]: Yeah, this was the first time I used it in a project. It really made sense to use it. Waiting for every thumbnail to load would be a bad ux

**Review by mrjvs — CHANGES_REQUESTED** [review]
Nice job, just a few things missing.
I'm gonna retarget the PR to the v4 branch as its on the feature list for that version.

If you need assistance with the requested changes, we can also merge the incomplete PR and then I will finish it up.

File: src/utils/thumbnailCreator.ts [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  video.crossOrigin = "anonymous";
+
+  // Wait for the video metadata to load
+  await new Promise((resolve, reject) => {
+    video.addEventListener("loadedmetadata", resolve);
+    video.addEventListener("error", reject);
+  });
+
+  canvas.height = video.videoHeight * SCALE_FACTOR;
+  canvas.width = video.videoWidth * SCALE_FACTOR;
```
[the remark this answers is not in the document]
Comment (frost768) [reply src/utils/thumbnailCreator.ts:25]: For this I will create an interface and implement it for hls and mp4 videos.
Hls one will have the same logic:
1. Create an instance of hls with the video source
2. Listen to corresponding metadata and seek events in hls js lib and seek to desired tims with hls controller
3. Convert fragment array data into base64 data url

Also an abortcontroller to cancel future requests

[the remark this answers is not in the document]
Comment (mrjvs) [reply src/utils/thumbnailCreator.ts:25]: sounds good. A tip for the aborting. if you make it an internal video component. (which are just components of the video player that are always present) you can use useEffect to detect when the videoplayer has unloaded. Its also a good place to keep a video element always loaded (and hidden). a bit more efficient than making a new video element everytime.

**Comment by mrjvs** [comment]
because the video player code will get a rewrite very soon for the new design, I'm gonna merge this now so the thumbnail component get roped in with the rewrite.

only thing missing was the design anyway, so good timing

**Review by mrjvs — APPROVED** [review]

**Pull request merged**